File: UN Training guide/Node 0/rabbitvoting-java/RabbitVotingTest.java
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class RabbitVotingTest {

    protected RabbitVoting solution;

    @Before
    public void setUp() {
        solution = new RabbitVoting();
    }

    @Test(timeout = 2000)
    public void testCase0() {
        String[] names = new String[]{"Alice", "Bill", "Carol", "Dick"};
        String[] votes = new String[]{"Bill", "Dick", "Alice", "Alice"};

        String expected = "Alice";
        String actual = solution.getWinner(names, votes);

        Assert.assertEquals(expected, actual);
    }

    @Test(timeout = 2000)
    public void testCase1() {
        String[] names = new String[]{"Alice", "Bill", "Carol", "Dick"};
        String[] votes = new String[]{"Carol", "Carol", "Bill", "Bill"};

        String expected = "";
        String actual = solution.getWinner(names, votes);

        Assert.assertEquals(expected, actual);
    }

    @Test(timeout = 2000)
    public void testCase2() {
        String[] names = new String[]{"Alice", "Bill", "Carol", "Dick"};
        String[] votes = new String[]{"Alice", "Alice", "Bill", "Bill"};

        String expected = "Bill";
        String actual = solution.getWinner(names, votes);

        Assert.assertEquals(expected, actual);
    }

    @Test(timeout = 2000)
    public void testCase3() {
        String[] names = new String[]{"Alice", "Bill"};
        String[] votes = new String[]{"Alice", "Bill"};

        String expected = "";
        String actual = solution.getWinner(names, votes);

        Assert.assertEquals(expected, actual);
    }

    @Test(timeout = 2000)
    public void testCase4() {
        String[] names = new String[]{"WhiteRabbit", "whiterabbit", "whiteRabbit", "Whiterabbit"};
        String[] votes = new String[]{"whiteRabbit", "whiteRabbit", "whiteRabbit", "WhiteRabbit"};

        String expected = "whiteRabbit";
        String actual = solution.getWinner(names, votes);

        Assert.assertEquals(expected, actual);
    }

}

File: un-training-guide/Node 0/rabbitvoting-c++/RabbitVotingTest.cpp
#include "RabbitVoting.h"
#include <iostream>
#include <string>
#include <vector>

using std::cerr;
using std::cout;
using std::endl;
using std::string;
using std::vector;

class RabbitVotingTest {

    static void assertEquals(int testCase, const string& expected, const string& actual) {
        if (expected == actual) {
            cout << "Test case " << testCase << " PASSED!" << endl;
        } else {
            cout << "Test case " << testCase << " FAILED! Expected: <" << '"' << expected << '"' << "> but was: <" << '"' << actual << '"' << '>' << endl;
        }
    }

    RabbitVoting solution;

    void testCase0() {
        string names_[] = {"Alice", "Bill", "Carol", "Dick"};
        vector<string> names(names_, names_ + (sizeof(names_) / sizeof(names_[0])));
        string votes_[] = {"Bill", "Dick", "Alice", "Alice"};
        vector<string> votes(votes_, votes_ + (sizeof(votes_) / sizeof(votes_[0])));
		string expected_ = "Alice";
        assertEquals(0, expected_, solution.getWinner(names, votes));
    }

    void testCase1() {
        string names_[] = {"Alice", "Bill", "Carol", "Dick"};
        vector<string> names(names_, names_ + (sizeof(names_) / sizeof(names_[0])));
        string votes_[] = {"Carol", "Carol", "Bill", "Bill"};
        vector<string> votes(votes_, votes_ + (sizeof(votes_) / sizeof(votes_[0])));
		string expected_ = "";
        assertEquals(1, expected_, solution.getWinner(names, votes));
    }

    void testCase2() {
        string names_[] = {"Alice", "Bill", "Carol", "Dick"};
        vector<string> names(names_, names_ + (sizeof(names_) / sizeof(names_[0])));
        string votes_[] = {"Alice", "Alice", "Bill", "Bill"};
        vector<string> votes(votes_, votes_ + (sizeof(votes_) / sizeof(votes_[0])));
		string expected_ = "Bill";
        assertEquals(2, expected_, solution.getWinner(names, votes));
    }

    void testCase3() {
        string names_[] = {"Alice", "Bill"};
        vector<string> names(names_, names_ + (sizeof(names_) / sizeof(names_[0])));
        string votes_[] = {"Alice", "Bill"};
        vector<string> votes(votes_, votes_ + (sizeof(votes_) / sizeof(votes_[0])));
		string expected_ = "";
        assertEquals(3, expected_, solution.getWinner(names, votes));
    }

    void testCase4() {
        string names_[] = {"WhiteRabbit", "whiterabbit", "whiteRabbit", "Whiterabbit"};
        vector<string> names(names_, names_ + (sizeof(names_) / sizeof(names_[0])));
        string votes_[] = {"whiteRabbit", "whiteRabbit", "whiteRabbit", "WhiteRabbit"};
        vector<string> votes(votes_, votes_ + (sizeof(votes_) / sizeof(votes_[0])));
		string expected_ = "whiteRabbit";
        assertEquals(4, expected_, solution.getWinner(names, votes));
    }

    public: void runTest(int testCase) {
        switch (testCase) {
            case (0): testCase0(); break;
            case (1): testCase1(); break;
            case (2): testCase2(); break;
            case (3): testCase3(); break;
            case (4): testCase4(); break;
            default: cerr << "No such test case: " << testCase << endl; break;
        }
    }

};

int main() {
    for (int i = 0; i < 5; i++) {
        RabbitVotingTest test;
        test.runTest(i);
    }
}
